inference_rules.adoc

```diff
@@ -821,6 +821,20 @@ The DLRMv2 MLPerf inference code has an option to aggregate multiple consecutive
 2. drawn uniformly from interval [`--samples-to-aggregate-min`, `--samples-to-aggregate-max`]	
 3. drawn from a custom distribution, with its quantile (inverse of CDP) specified in `--samples-to-aggregate-quantile-file=./tools/dist_quantile.txt`.
 
+=== LLM Benchmarks
+
+Q: What algorithm is used for the auto-regressive decoding loop?
+
+A: The benchmark uses the beam search algorithm described at a high level here: https://huggingface.co/blog/how-to-generate#beam-search. Specifically, we use a beam width of 4 and enable early termination. TODO(thomas,bhadri): Add more details here of the beam search in GPT-J.
+
+Q: MLPerf disallows caching queries. Is using a KV-cache in decoding allowed?
+
+A: Using a KV-cache is allowed in the same way as included in the reference model where it does not apply across queries. A KV-cache row is a tensor that must deal with the execution of a single inference query. Every input query must be computed in its entirety. Entries of the KV-cache must be treated similar to activations of a forward pass. They can be quantized with respect to the quantization rules. Similarly, they cannot be pruned (or sparsified).
```

Comment:
There was concern regarding the "cannot be pruned (or sparsified)" part of this in the inference working group today. Anton brought up a valid point that some implementations might choose to cache some entries while rematerializing some K/V vectors. I think that is a valid scenario that we want to allow. In my mind, we were not trying to prohibit that with the above rule.

Reply:
Thanks for the note Akhil. Yes, I agree. We want to ensure with the rules that we standardize using the KV-cache as it has a lot of moving parts. However, not using a KV cache or rematerializing some of the values should be allowed. Pruning and sparsification changes model predictions and are disallowed in general as part of model equivalence rules. Not using KV-cache isn't pruning as long as the values are rematerialized.